doc/tests-updates.md

```diff
@@ -0,0 +1,381 @@
+
+- Request-response chain integrity. Need to validate that every client receives
+its responses in the correct order.
+
+- Fault injections. Based on other types of tests, but random errors to be
+emulated using SystemTap or eBPF.
+
+- Fussing. Generating random requests and responses, Tempesta must stay
+operational and serve as many requests as possible.
+
+- Workload testing. Generate pretty high request rate to confirm data safety
+on highly concurrent code execution. The more CPUs is used for Tempesta -
+the better. But CPU utilisation must be less than 80% to assure that no extreme
+conditions happen.
+
+- Stress testing. Testing under extreme conditions, when request rate is higher
+than Tempasta can serve.
+
+- Low level TCP testing. Some features can affect TCP behaviour, so need to
+validate that the Tempesta conforms RFCs and doesn't break kernel behaviour.
+
+- TLS testing. Having in-house TLS implementation gives great powers and great
+responsibility. We must assure TLS support messages integrity, encryption
+integrity. It differs from other tests, since it's important to work with raw
+data lying under HTTP on client and server side.
+
+- Tempesta-originated requests. Pretty specific tests, since Tempesta may
+generate requests on its own, e.g. background cache revalidation, or health
+monitor requests. It differs from other tests, since it's important to validate
+support TCP messages.
+
+
+There also some requirements on tests process:
+
+- Full, reduced and extra-stability set of tests. Reduced for developers to
+run before making pull requests. Full to run against pull requests.
+Extra-stability for every night testing to assure stability under unpredictable
+loads.
+
+- Easy new test addition.
+
+- Easy test case failure analysis.
+
+- Generate a few tests based on singe test description. E.g. in fussing test
+it's required to know, how the Tempesta would react on a message, and what
+should be threated as error. The 'splitting on different skbs' is actually
+fussing test feature, not functional. Usually it also make sense to generate
+workload test based on functional profile.
+
+
+## Problems of current implementation
+
+The main problem is having two frameworks in the same time. Old, feature rich,
+but too sophisticated, with requirement to understand 5-6 files to understand
+what the test is actually does. And a new, more friendly, but not complete and
+with lack of the most important features. Transition of an old one to a new
+wasn't completed, and the development of both was rather unorganized, thus
+new framework is becoming more and more complex and loses its advantages.
+
+In both frameworks every test type is implemented as unique test, so it's
+required to write message integrity, fussing, workload and many other test
+cases for the same situation.
+
+Heavy dependency between tests. When a new feature is added, it's added to the
+test framework core and affects a lot of tests. Sometimes unpredictably and
+badly.
```

Comment:
Yes, we had them. We need to discuss this in chat considering particular cases. There are different possibilities what happens when we're in the situation - we grow the framework that it can run many different things in future or we just keep case-specific login in the framework. Let's talk on the cases in the chat. The cases should make the framework design problems clear.